### ISSUE TITLE
CoreDNS eats up CPU and I/O

### ISSUE THREAD

**Issue by MartinEmrich** [coredns/coredns#2660]
Hi!

CoreDNS in my K8s cluster (3 Masters, 3 Workers, maybe 70 pods in all) is eating up all CPU (200-300%) and has immense I/O as seen with iotop:

```
Total DISK READ :     213.67 M/s | Total DISK WRITE :     156.35 K/s
Actual DISK READ:     214.11 M/s | Actual DISK WRITE:     159.83 K/s
    TID  PRIO  USER     DISK READ  DISK WRITE  SWAPIN     IO>    COMMAND                                                                                                                                                                                                       
 842404 be/4 root	48.65 M/s    0.00 B/s  0.00 % 24.67 % coredns -conf /etc/coredns/Corefile
 850515 be/4 root	35.77 M/s    0.00 B/s  0.00 % 21.27 % coredns -conf /etc/coredns/Corefile
 842289 be/4 root	47.30 M/s    0.00 B/s  0.00 % 21.12 % coredns -conf /etc/coredns/Corefile
 842420 be/4 root       45.04 M/s    0.00 B/s  0.00 % 20.85 % coredns -conf /etc/coredns/Corefile
 842413 be/4 root	12.47 M/s    0.00 B/s  0.00 % 18.88 % coredns -conf /etc/coredns/Corefile
 842435 be/4 root	16.34 M/s    0.00 B/s  0.00 % 12.10 % coredns -conf /etc/coredns/Corefile
```

K8s Version 1.13.0
CoreDNS version 1.2.6

CoreDNS Config (Default from kubeadm/kubernetes):
```
    .:53 {
        errors
        health
        kubernetes cluster.local in-addr.arpa ip6.arpa {
           pods insecure
           upstream
           fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        proxy . /etc/resolv.conf
        cache 30
        loop
        reload
        loadbalance
    }
```

Host resolv.conf:
```
search some.sub.domain.company.com
nameserver 10.12.96.3
nameserver 172.24.187.220
nameserver 172.24.186.30
```

`kubectl logs` does not give any clues.

What could be wrong?

**Comment by johnbelamaric** [comment]
No clue. It should be doing very very low reads (just the reloads of the config and resolv.conf periodically). Anything from `kubectl describe`?

**Comment by MartinEmrich** [comment]
Sorry, nothing interesting there:

```
Name:               coredns-6f685fffbf-2z9sx
Namespace:          kube-system
Priority:           0
PriorityClassName:  <none>
Node:               kl-k8s-dev-worker01/10.12.102.82
Start Time:         Thu, 07 Mar 2019 13:34:56 +0100
Labels:             k8s-app=kube-dns
                    pod-template-hash=6f685fffbf
Annotations:        <none>
Status:             Running
IP:                 10.32.0.7
Controlled By:      ReplicaSet/coredns-6f685fffbf
Containers:
  coredns:
    Container ID:  docker://8bd10e9dfe6b7f3f397a992f170232d52c24f8da62747eaeaeecf26677444525
    Image:         k8s.gcr.io/coredns:1.2.6
    Image ID:      docker-pullable://k8s.gcr.io/coredns@sha256:81936728011c0df9404cb70b95c17bbc8af922ec9a70d0561a5d01fefa6ffa51
    Ports:         53/UDP, 53/TCP, 9153/TCP
    Host Ports:    0/UDP, 0/TCP, 0/TCP
    Args:
      -conf
      /etc/coredns/Corefile
    State:          Running
      Started:      Fri, 08 Mar 2019 10:11:52 +0100
    Last State:     Terminated
      Reason:       Error
      Exit Code:    255
      Started:      Thu, 07 Mar 2019 13:35:00 +0100
      Finished:     Fri, 08 Mar 2019 10:10:51 +0100
    Ready:          True
    Restart Count:  1
    Limits:
      memory:  170Mi
    Requests:
      cpu:        100m
      memory:     70Mi
    Liveness:     http-get http://:8080/health delay=60s timeout=5s period=10s #success=1 #failure=5
    Environment:  <none>
    Mounts:
      /etc/coredns from config-volume (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from coredns-token-tg8qj (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  config-volume:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      coredns
    Optional:  false
  coredns-token-tg8qj:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  coredns-token-tg8qj
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     CriticalAddonsOnly
                 node-role.kubernetes.io/master:NoSchedule
                 node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:          <none>
```

**Comment by johnbelamaric** [comment]
Can you try setting `resyncperiod 0s`? That will disable the client-go API resync, which may be consuming CPU and possibly network I/O.

**Comment by MartinEmrich** [comment]
@johnbelamaric I did set the option, and stared at `iotop`for some time, and saw no I/O... Seems to be working.

**Comment by johnbelamaric** [comment]
Ok. How's the CPU? Does iotop differentiate network and disk I/O (above says "disk")?

**Comment by MartinEmrich** [comment]
iotop only displays disk I/O, that looks fine now.

CPU usage (with regular `top` or our monitoring) looks good, too, and seems to be unaffected by the change.

**Comment by johnbelamaric** [comment]
Huh. I wouldn't expect this change to affect disk I/O, only network I/O and CPU.

**Comment by MartinEmrich** [comment]
Mabye it was just coincidence... All of this smells strange, I'd wait until maybe someone else has the same problem.

**Comment by miekg** [comment]
closing. Re-open if you see it again or have a root-cause

**Comment by kzidane** [comment]
I'm experiencing the same problem on EKS using `602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/coredns:v1.2.2`. Here are the relevant parts from `top` and `iotop` outputs:

```
$ iotop
Total DISK READ :     127.43 M/s | Total DISK WRITE :       6.50 M/s
Actual DISK READ:     127.11 M/s | Actual DISK WRITE:       7.67 M/s
  TID  PRIO  USER     DISK READ  DISK WRITE  SWAPIN     IO>    COMMAND
21914 be/4 root        7.07 M/s    0.00 B/s  0.00 % 95.26 % coredns -conf /etc/coredns/Corefile
21910 be/4 root        7.56 M/s    0.00 B/s  0.00 % 95.17 % coredns -conf /etc/coredns/Corefile
21896 be/4 root        7.89 M/s    0.00 B/s  0.00 % 95.09 % coredns -conf /etc/coredns/Corefile
 8684 be/4 root        7.51 M/s    0.00 B/s  0.00 % 94.91 % coredns -conf /etc/coredns/Corefile
21916 be/4 root        7.70 M/s    0.00 B/s  0.00 % 94.60 % coredns -conf /etc/coredns/Corefile
25860 be/4 root        6.43 M/s    0.00 B/s  0.00 % 94.56 % coredns -conf /etc/coredns/Corefile
21889 be/4 root        7.61 M/s    0.00 B/s  0.00 % 94.49 % coredns -conf /etc/coredns/Corefile
21943 be/4 root        9.51 M/s    0.00 B/s  0.00 % 94.30 % coredns -conf /etc/coredns/Corefile
21904 be/4 root        7.76 M/s    0.00 B/s  0.00 % 93.92 % coredns -conf /etc/coredns/Corefile
21899 be/4 root        7.76 M/s    0.00 B/s  0.00 % 93.88 % coredns -conf /etc/coredns/Corefile
21939 be/4 root        6.27 M/s    0.00 B/s  0.00 % 92.54 % coredns -conf /etc/coredns/Corefile
30102 be/4 root        8.37 M/s    0.00 B/s  0.00 % 91.31 % coredns -conf /etc/coredns/Corefile
21905 be/4 root        8.61 M/s    0.00 B/s  0.00 % 90.58 % coredns -conf /etc/coredns/Corefile
21891 be/4 root        7.44 M/s    0.00 B/s  0.00 % 83.03 % coredns -conf /etc/coredns/Corefile
21901 be/4 root        7.03 M/s    0.00 B/s  0.00 % 60.28 % coredns -conf /etc/coredns/Corefile
21909 be/4 root        4.81 M/s    0.00 B/s  0.00 % 51.89 % coredns -conf /etc/coredns/Corefile
 6949 be/4 root        4.07 M/s    0.00 B/s  0.00 % 42.04 % coredns -conf /etc/coredns/Corefile
21911 be/4 root        2.35 M/s    0.00 B/s  0.00 % 32.67 % coredns -conf /etc/coredns/Corefile
22533 be/4 root      861.31 K/s    0.00 B/s  0.00 %  8.36 % coredns -conf /etc/coredns/Corefile
21903 be/4 root      272.38 K/s    0.00 B/s  0.00 %  2.85 % coredns -conf /etc/coredns/Corefile
26728 be/4 root      265.02 K/s    0.00 B/s  0.00 %  2.36 % coredns -conf /etc/coredns/Corefile
21915 be/4 root      346.00 K/s    0.00 B/s  0.00 %  2.03 % coredns -conf /etc/coredns/Corefile
```

```
$ top 
top - 18:21:14 up 205 days, 20:36,  1 user,  load average: 15.72, 11.03, 8.91
Tasks: 638 total,   1 running, 487 sleeping,   0 stopped,   0 zombie
%Cpu(s):  2.1 us,  8.2 sy,  0.0 ni, 48.2 id, 41.4 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem : 31889996 total,  1510604 free,  7442264 used, 22937128 buff/cache
KiB Swap:        0 total,        0 free,        0 used. 21977644 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
21850 root      20   0  206416 164652      0 S  93.8  0.5  24:35.58 coredns
```

```
$ kubectl version --short
Server Version: v1.12.10-eks-aae39f
```

This keeps going for a few minutes then things get back to normal. Unfortunately I'm not sure what's causing this or how to reproduce. Any thoughts on how to track down and fix? Thank you!

**Comment by chrisohaver** [comment]
> Any thoughts on how to track down ...

Do you collect coredns metrics?  Did coredns receive more query load during this time?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] CoreDNS eats ..." ]
>> Any thoughts on how to track down ...
>
>Do you collect coredns metrics?  Did coredns receive more query load during this time?

Can it be a loop? The *loop* plugin only checks on startup...

**Comment by kzidane** [comment]
Sorry for the slow reply! I did notice a bump in the DNS request count around the time this happened. I didn't have enough metrics data to check if that was the case too the few times I experienced this issue before this time, but I'll keep an eye in the future. I ended up scaling up the coredns deployment and I haven't seen this behavior since. I'll let you know if I run into it again. Thanks so much!
